### DISCUSSION TITLE
How to I get do the initialization synchronously?

### DISCUSSION THREAD

**Discussion by aswinfrancis91** [i18next/i18next#1547] [Q&A]
I am using i81next with chained backend, localStorage backend and http backend. I wrote a wrapper for i18next's init() and t() as suggested [here](https://phrase.com/blog/posts/javascript-i18n-i18next-moment-js/)
 Mine is an MVC application and so in shared layout I refer the app.js where I set 'window.lang' which can call init() and window.translate which can call t(). I set it t window so that I can use it anywhere within the site.

Now in the same shared layout I call window.lang.init() which internally calls i18next.init() with the necessary option. Now in other partial views I add extra namespaces as needed in that page using loadNamespaces and then call window.translate(key) which internally calls i18next.t(key). 

The problem is that at times it will come as undefined. If I enable debug, in console I can see that its attempting to translate before initialization happens even though when I checked the flow, initialization is called first. Is there a way to let the initialization finish before processing anything else?
![image](https://user-images.githubusercontent.com/1233653/107019898-3bd70700-67c8-11eb-9dd7-8109c2a9e3a5.png)

Once the page is fully loaded if I call window.translate(key that failed above') it will work fine this time

**Comment by jamuhl** [comment]
It says wait for the callback or Promise -> init, changeLanguage....are asyc

**Reply by aswinfrancis91** [reply]
I apologize if my understanding is completely wrong.
I have written the below function to initialize i18next and moment locale
![image](https://user-images.githubusercontent.com/1233653/107035007-88790d00-67dd-11eb-9e79-efc718beecb0.png)
I am calling it in a cshtml file like 
![image](https://user-images.githubusercontent.com/1233653/107035081-9f1f6400-67dd-11eb-90fc-242b4785c2c2.png)
I have a translate function to call i18next.t()
![image](https://user-images.githubusercontent.com/1233653/107035125-af374380-67dd-11eb-9e7b-44669f251111.png)
Now in the cshtml page where i want translation , i call
window.translate("key")

**Reply by jamuhl** [reply]
nope...that is not correct - your init call is not async. You call one thing async in it...that is all...stackoverflow might be a good please to find help

**Reply by aswinfrancis91** [reply]
So I made my init call async
```
//Wrapper for i18next's init()
const initI18next = function (opt) {
    isInit = date.init(opt.lng).then(function () {
        i18next.use(Backend).init(opt);
    });
    return isInit;
}
```
And then my loadspace also sync
```
//Wrapper for i18Next's loadNamespace call. Add addition namespace not in init()
const addNewNamespaces = function (namespaces) {
    isLoaded = i18next.loadNamespaces(namespaces);
    return isLoaded;
}
```
My translate wrapper
```
translate: async function (key, opt) {
        await Promise.all([isInit, isLoaded]).then(() => {
            return i18next.t(key, opt);
        });
    }
```
But in html/js, the object would be Promsie as expected. I really just want to ensure that t() would wait for init and loadnamespace before attempting translation. Is there no way to make the call sync. What about the flag 'initImmediate'? I am using 'i18next-http-backend'

**Reply by jamuhl** [reply]
loadNamespace is async too, t is sync...

I can't help you with your code...more than just saying handle async functions correctly
